autopush_rs/src/client.rs

```diff
@@ -0,0 +1,594 @@
+                    return Ok(next_state.into())
+                }
+                match try_ready!(self.data.input()) {
+                    ClientMessage::Register { channel_id, key } => {
+                        self.data.process_register(channel_id, key)
+                    },
+                    ClientMessage::Unregister { channel_id, code } => {
+                        self.data.process_unregister(channel_id, code)
+                    },
+                    ClientMessage::Ack { updates } => {
+                        self.data.process_acks(updates)
+                    }
+                    _ => return Err("Invalid state transition".into())
+                }
+            },
+            ClientState::WaitingForDelete(ref mut response) => {
+                debug!("State: WaitingForDelete");
+                try_ready!(response.poll());
+                ClientState::WaitingForAcks
+            },
+            ClientState::WaitingForDropUser(ref mut response) => {
+                debug!("State: WaitingForDropUser");
+                try_ready!(response.poll());
+                ClientState::Done
+            },
+            ClientState::Await => {
+                debug!("State: Await");
+                if self.data.webpush.as_ref().unwrap().flags.check {
+                    return Ok(ClientState::CheckStorage.into())
+                }
+                match try_ready!(self.data.input_or_notif()) {
+                    Either::A(ClientMessage::Register { channel_id, key }) => {
+                        self.data.process_register(channel_id, key)
+                    },
+                    Either::A(ClientMessage::Unregister { channel_id, code }) => {
+                        self.data.process_unregister(channel_id, code)
+                    },
+                    Either::B(ServerNotification::Notification(notif)) => {
+                        let webpush = self.data.webpush.as_mut().unwrap();
+                        webpush.unacked_direct_notifs.push(notif.clone());
+                        debug!("Got a notification to send, sending!");
+                        ClientState::FinishSend(
+                            Some(ServerMessage::Notification(notif)),
+                            Some(Box::new(ClientState::WaitingForAcks)),
+                        )
+                    },
+                    Either::B(ServerNotification::CheckStorage) => {
+                        let webpush = self.data.webpush.as_mut().unwrap();
+                        webpush.flags.include_topic = true;
+                        webpush.flags.check = true;
+                        ClientState::Await
+                    },
+                    _ => return Err("Invalid message".into())
+                }
+            },
+            ClientState::ShutdownCleanup(ref mut err) => {
+                debug!("State: ShutdownCleanup");
+                if let Some(err_obj) = err.take() {
+                    debug!("Error for shutdown: {}", err_obj);
```

Comment:
Errors from error chain also have "backtraces" associated with them which should contain extra error context ins ome situations, so maybe this'll want a helper method to stringify the error plus the entire error backtrace associated with it?

Reply:
Oh, how does one get that?

Reply:
I think it'd look [something like this](https://github.com/mozilla-services/autopush/blob/5bf2de6403e3f7a5ff8b4910b86d06c99b023cf7/autopush_rs/src/server.rs#L310-L317)